kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLog.java

```diff
@@ -102,25 +107,71 @@ public class PartitionLog {
     private final Time time;
     private final TopicPartition topicPartition;
     private final String fullPartitionName;
-    private final EntryFormatter entryFormatter;
+    private AtomicReference<CompletableFuture<EntryFormatter>> entryFormatter = new AtomicReference<>();
     private final ProducerStateManager producerStateManager;
 
+    private final ImmutableMap<String, EntryFilterWithClassLoader> entryfilterMap;
+
     public PartitionLog(KafkaServiceConfiguration kafkaConfig,
                         RequestStats requestStats,
                         Time time,
                         TopicPartition topicPartition,
                         String fullPartitionName,
-                        EntryFormatter entryFormatter,
+                        ImmutableMap<String, EntryFilterWithClassLoader> entryfilterMap,
                         ProducerStateManager producerStateManager) {
         this.kafkaConfig = kafkaConfig;
+        this.entryfilterMap = entryfilterMap;
         this.requestStats = requestStats;
         this.time = time;
         this.topicPartition = topicPartition;
         this.fullPartitionName = fullPartitionName;
-        this.entryFormatter = entryFormatter;
         this.producerStateManager = producerStateManager;
     }
 
+    private CompletableFuture<EntryFormatter> getEntryFormatter(
+            CompletableFuture<Optional<PersistentTopic>> topicFuture) {
+        return entryFormatter.accumulateAndGet(null, (current, ___) -> {
+            if (current != null) {
+                return current;
+            }
+            return topicFuture.thenCompose((persistentTopic) -> {
+                TopicName logicalName = TopicName.get(persistentTopic.get().getName());
+                TopicName actualName;
+                if (logicalName.isPartitioned()) {
+                    actualName = TopicName.getPartitionedTopicName(persistentTopic.get().getName());
+                } else {
+                    actualName = logicalName;
+                }
+                CompletableFuture<EntryFormatter> result = persistentTopic.get().getBrokerService()
+                        .fetchPartitionedTopicMetadataAsync(actualName)
+                        .thenApply(metadata -> {
+                            if (metadata.partitions > 0) {
+                                return buildEntryFormatter(metadata.properties);
+                            } else {
+                                return buildEntryFormatter(persistentTopic.get().getManagedLedger().getProperties());
+                            }
+                        });
+                result.exceptionally(ex -> {
+                   log.error("Cannot create the EntryFormatter for {}", fullPartitionName, ex);
+                   entryFormatter.set(null);
```

Comment:
Why set null here?

Reply:
Because if there is an error the entryFormatter CompletableFuture will stay as a "failed future" and the PartitionLog will never be able to work.

Maybe I can add a test case for this (it is pretty hard, but with some mockito maybe it is doable)

Reply:
I am not sue it is worth to add a test case.
I added a comment that explains that this code is running in a separate thread